### PR TITLE
Add DVB subtitle support

### PR DESCRIPTION
**Pull request by tresvecesseis**
This is a first patch to bring DVB subtitle support to Exoplayer, it is working we SD and HD subtitle format. DVB subs is the standard subtitling format in the countries that use the DVB-t/s/c for TV broadcasting and the IP providers based the DVB-IPI standards. We have implemented a DvbSubtitlesTrackRenderer() but we could change it to a generic GraphicTextTrackRenderer() if you want to have other graphic subtitle formats as the DVD VOBSUBs .

### CODE REVIEW

File: library/src/main/java/com/google/android/exoplayer/text/DVBSubs/DvbSubtitlesParser.java [hunk 1]
```diff
@@ -0,0 +1,922 @@
...
+  private boolean wordAligned() {
+    int bytePos = (tsStream.getPosition() - 1)/8;
+    if (bytePos % 2 != 0){
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  public DvbSubtitlesParser   (Paint paint, Bitmap bitmap) {
```
Comment (ened) [inline library/src/main/java/com/google/android/exoplayer/text/DVBSubs/DvbSubtitlesParser.java]: Extra space?

**Comment by cooliobr** [comment]
Hi, one question, this commit are working with DVBSUB ?
i try some streams but not work

**Comment by tresvecesseis** [comment]
Yep we use it, but the patch is old, i will try to rebase it to the current
code base.

**Comment by cooliobr** [comment]
please, if you update let me know, and if possible try this stream
http://200.194.237.226:8182/live/fox1/playlist.m3u8?x1=test&x2=test&canal=FOX1&mac=cc:2d:8c:aa:02:c2&ipaddr=187.58.37.123

**Comment by tresvecesseis** [comment]
@cooliobr your stream have two subs tracks, one DVB and other SCTE, both are correctly signalled and the DVB one should be decoded, tomorrow when I have access to an android device I will test your stream

**Comment by cooliobr** [comment]
Thanks, i will appreciate your help.

**Comment by cooliobr** [comment]
Hi, any news to me ?

**Comment by tresvecesseis** [comment]
We hace just finished a patch to be able to render your subs, I will push
it soon

**Comment by tresvecesseis** [comment]
This patch should do the trick, The TsExtractor part solves an issue with the audio language parsing, and  the DvbSubtitlesParse solves the subtitile visualization problema, we can´t connect to your HLS stream but the dumped ts segments work propelry:

 

diff --git a/library/src/main/java/com/google/android/exoplayer/extractor/ts/TsExtractor.java b/library/src/main/java/com/google/android/exoplayer/extractor/ts/TsExtractor.java

diff --git a/library/src/main/java/com/google/android/exoplayer/text/dvbsubs/DvbSubtitlesParser.java b/library/src/main/java/com/google/android/exoplayer/text/dvbsubs/DvbSubtitlesParser.java

index 25b997f..ea1f1b6 100644

--- a/library/src/main/java/com/google/android/exoplayer/extractor/ts/TsExtractor.java

+++ b/library/src/main/java/com/google/android/exoplayer/extractor/ts/TsExtractor.java

@@ -347,7 +347,7 @@ public final class TsExtractor implements Extractor {

           // Read descriptors in PES packets containing private data.

           streamType = readPrivateDataStreamType(sectionData, esInfoLength);

         } else {

-          if (streamType == TS_STREAM_TYPE_MPA || streamType == TS_STREAM_TYPE_MPA_LSF) {

+          if ((streamType == TS_STREAM_TYPE_MPA || streamType == TS_STREAM_TYPE_MPA_LSF) && esInfoLength != 0) {

             descriptorScratch = new ParsableBitArray(new byte[esInfoLength]);

             sectionData.readBytes(descriptorScratch, esInfoLength);

           } else {

@@ -364,21 +364,25 @@ public final class TsExtractor implements Extractor {

         int type = -1;

         switch (streamType) {

           case TS_STREAM_TYPE_MPA:

-            type = descriptorScratch.readBits(8);

-            switch (type) {

-              case 0x0a:

-                language = readDescriptorAudio(descriptorScratch);

-                break;

+            if (esInfoLength > 0) {

+              type = descriptorScratch.readBits(8);

+              switch (type) {

+                case 0x0a:

+                  language = readDescriptorAudio(descriptorScratch);

+                  break;

+              }

             }

             pesPayloadReader = new MpegAudioReader(output.track(elementaryPid), language);

             pidToTypeMap.put(elementaryPid, TS_STREAM_TYPE_MPA);

             break;

           case TS_STREAM_TYPE_MPA_LSF:

-            type = descriptorScratch.readBits(8);

-            switch (type) {

-              case 0x0a:

-                language = readDescriptorAudio(descriptorScratch);

-                break;

+            if (esInfoLength > 0) {

+              type = descriptorScratch.readBits(8);

+              switch (type) {

+                case 0x0a:

+                  language = readDescriptorAudio(descriptorScratch);

+                  break;

+              }

             }

             pesPayloadReader = new MpegAudioReader(output.track(elementaryPid), language);

             pidToTypeMap.put(elementaryPid, TS_STREAM_TYPE_MPA_LSF);

diff --git a/library/src/main/java/com/google/android/exoplayer/text/dvbsubs/DvbSubtitlesParser.java b/library/src/main/java/com/google/android/exoplayer/text/dvbsubs/DvbSubtitlesParser.java

index 6084f7d..af47b26 100644

--- a/library/src/main/java/com/google/android/exoplayer/text/dvbsubs/DvbSubtitlesParser.java

+++ b/library/src/main/java/com/google/android/exoplayer/text/dvbsubs/DvbSubtitlesParser.java

@@ -264,7 +264,7 @@ public class DvbSubtitlesParser {

       if(displayCounter == 0) {

         displays = new DvbSubDisplay[1];

         displays[0] = new DvbSubDisplay();

-        displayCounter = 0;

+        displayCounter = 1;

       } else {

         displays = Arrays.copyOf(displays, displayCounter + 1);

         displayCounter++;

--

2.8.1

**Comment by cooliobr** [comment]
is possible you put this diff in your repository ?
https://github.com/tresvecesseis/ExoPlayer will be better for me



Att.

Rafael Machado
REGS IT LTDA.
ANTIK Business Representative BRASIL
"Believe in the future, believe in your success"
"Não existe patch para a estupidez humana"
"Nós somos o que pensamos (só não pense que você é um super-herói e não
tente voar)"

2016-05-31 7:52 GMT-03:00 tresvecesseis <notifications@github.com>:

**Comment by tresvecesseis** [comment]
It will take a while, I´m rebasing all our trees  to the current dev tree, sadly I didn’t began with the DVB subs one

 

Sergio

**Comment by cooliobr** [comment]
how i can do checkout of this source ?
sorry if i do many questions but, i will appreciate your help



Att.

Rafael Machado
REGS IT LTDA.
ANTIK Business Representative BRASIL
"Believe in the future, believe in your success"
"Não existe patch para a estupidez humana"
"Nós somos o que pensamos (só não pense que você é um super-herói e não
tente voar)"

2016-05-31 13:29 GMT-03:00 tresvecesseis <notifications@github.com>:

**Comment by tresvecesseis** [comment]
Our tres are private and based in an old Exoplayer versión, I’ll update my github when I complete the update of the trees to the current dev tree, It could take a couple of days, maybe less.

 

Sergio

**Comment by cooliobr** [comment]
ok, thanks for your help :D
if you need stream channels from Brasil i can give you



Att.

Rafael Machado
REGS IT LTDA.
ANTIK Business Representative BRASIL
"Believe in the future, believe in your success"
"Não existe patch para a estupidez humana"
"Nós somos o que pensamos (só não pense que você é um super-herói e não
tente voar)"

2016-05-31 13:50 GMT-03:00 tresvecesseis <notifications@github.com>:

**Comment by tresvecesseis** [comment]
How did you configure the exoplayer library to process your streams?, we were unable to do that and we had to dump your TSs to be able to debug the problem.

 

Sergio

**Comment by tresvecesseis** [comment]
I just updated our github with a current versión of the dvbsubs tree:

 

https://github.com/tresvecesseis/ExoPlayer/tree/dev-dvbsubs

 

Sergio

**Comment by tresvecesseis** [comment]
Could you test it?

 

From: "tresvecesseis@gmail.com" <tresvecesseis@gmail.com>
Date: Tuesday 31 May 2016 at 19:33
To: google/ExoPlayer <reply@reply.github.com>, google/ExoPlayer <ExoPlayer@noreply.github.com>
Cc: Comment <comment@noreply.github.com>
Subject: Re: [google/ExoPlayer] Add DVB subtitle support (#705)

 

How did you configure the exoplayer library to process your streams?, we were unable to do that and we had to dump your TSs to be able to debug the problem.

 

Sergio

**Pull request closed without merging**